### PR TITLE
Singularize!

### PR DESCRIPTION
**Pull request by o-smirnov**
OK I have most of a real-life recipe running with singularity containers, so in the interest of release early, release often, let's push this out.

I'll open a cult-cargo PR to accompany this shortly.

Putting this at the start of a recipe enables the singularity backend. I haven't yet updated the click options to set this from the command line, but we should.

```yml
opts:
  backend:
    singularity:
      enable: true
      auto_update: true
```

### CODE REVIEW

File: stimela/backends/__init__.py [hunk 1]
```diff
@@ -1,32 +1,91 @@
...
+from typing import Union, Dict, Any, List, Optional
+from enum import Enum
+from omegaconf import OmegaConf, MISSING
 
+import stimela.kitchen
 from stimela.exceptions import SchemaError
-from scabha.basetypes import File, Directory, MS
-
+from scabha.basetypes import File, Directory, MS, EmptyDictDefault, EmptyListDefault
+
```
Comment (sphemakh) [inline stimela/backends/__init__.py]: Why leave this commented-out code here?

File: stimela/backends/__init__.py [hunk 2]
```diff
@@ -1,32 +1,91 @@
...
 @dataclass
-class StimelaImageBuildInfo:
-    stimela_version: str = ""
-    user: str = ""
-    date: str = ""
-    host: str = ""  
+class StimelaImage:
+    name: str
+    version: str
+    # name: str = MISSING
```
Comment (sphemakh) [inline stimela/backends/__init__.py]: ditto

File: stimela/backends/kubernetes/__init__.py [hunk 3]
```diff
@@ -1,3 +1,11 @@
+from typing import Dict, List, Optional, Any
+from enum import Enum
+from omegaconf import OmegaConf
+from dataclasses import dataclass
+from scabha.basetypes import EmptyDictDefault, EmptyListDefault
+import json, yaml
```
Comment (sphemakh) [inline stimela/backends/kubernetes/__init__.py:6]: one import per line

File: stimela/backends/kubernetes/run_kube.py [hunk 4]
```diff
@@ -2,16 +2,13 @@
 import logging, time, json, datetime, yaml, os.path, uuid, pathlib
```
Comment (sphemakh) [inline stimela/backends/kubernetes/run_kube.py:2]: One import per line

File: stimela/backends/native/__init__.py [hunk 5]
```diff
@@ -1,4 +1,7 @@
...
-STATUS = "ok"
+def is_available():
+    return True
 
-from .run_native import run
\ No newline at end of file
+def get_status():
+    return "OK"
+
+from .run_native import run, build_command_line, update_rlimits
```
Comment (sphemakh) [inline stimela/backends/native/__init__.py]: add new line at end file

File: stimela/config.py [hunk 6]
```diff
@@ -206,22 +161,27 @@ class StimelaLibrary(object):
...
         lib: StimelaLibrary = StimelaLibrary()
         cabs: Dict[str, Cab] = EmptyDictDefault()
         opts: StimelaOptions = StimelaOptions()
         vars: Dict[str, Any] = EmptyDictDefault()
         run:  Dict[str, Any] = EmptyDictDefault()
 
-    base_configs_glob = f"{STIMELA_DIR}/cargo/base/*/*.yaml"
-    lib_configs_glob = f"{STIMELA_DIR}/cargo/lib/params/*.yaml"
-    cab_configs_glob = f"{STIMELA_DIR}/cargo/cab/*.yaml"
+    ## replaced by cult-cargo
```
Comment (sphemakh) [inline stimela/config.py:169]: remove old code, don't comment it out

File: stimela/config.py [hunk 7]
```diff
@@ -309,7 +258,7 @@ def _load(conf, config_file):
         if not CONFIG_LOADED:
             log.info("no user-supplied configuration files given, using defaults")
 
-        dependencies.replace((base_configs_glob, cab_configs_glob, lib_configs_glob), STIMELA_DIR)
+        # dependencies.replace((base_configs_glob, cab_configs_glob, lib_configs_glob), STIMELA_DIR)
```
Comment (sphemakh) [inline stimela/config.py:255]: remove

File: stimela/kitchen/step.py [hunk 8]
```diff
@@ -66,8 +65,8 @@ class Step:
     assign_based_on: Dict[str, Any] = EmptyDictDefault()
                                                     # assigns recipe-level variables when step is executed based on value of another variable
 
-    # runtime settings
-    runtime: Dict[str, Any] = EmptyDictDefault()
+    # optional backend settings
+    backend: Optional[Dict[str, Any]] = None
```
Comment (sphemakh) [inline stimela/kitchen/step.py:70]: ditto for old code

File: stimela/main.py [hunk 9]
```diff
@@ -40,8 +40,8 @@ def resolve_command(self, ctx, args):
 
 
 @click.group(cls=RunExecGroup)
-@click.option('--backend', '-b', type=click.Choice(config.Backend._member_names_), 
-                help="Backend to use (for containerization).")
+# @click.option('--backend', '-b', type=click.Choice(config.Backend._member_names_),
```
Comment (sphemakh) [inline stimela/main.py]: ditoo for old code

File: stimela/main.py [hunk 10]
```diff
@@ -96,20 +96,20 @@ def cli(backend, config_files=[], config_dotlist=[], include=[], verbose=False,
         stimelogging.update_file_logger(log, stimela.CONFIG.opts.log, nesting=-1, subst=subst)
 
 
-    # print backends
-    log.info(f"available backends: {' '.join(config.AVAILABLE_BACKENDS)}")
+    # # print backends
```
Comment (sphemakh) [inline stimela/main.py]: Ditto

File: stimela/stimela.conf [hunk 11]
```diff
@@ -0,0 +1,3 @@
+images:
+    default-python: quay.io/stimela2/python-astro:cc0.0.2
+    default-casa: quay.io/stimela2/casa:cc0.0.2
```
Comment (sphemakh) [inline stimela/stimela.conf]: add new line

**Review by SpheMakh — CHANGES_REQUESTED** [review]

**Comment by sphemakh** [comment]
Run auto [pep8](https://pypi.org/project/autopep8) and re-commit

File: stimela/main.py (hunk 9 above)
Comment (o-smirnov) [reply stimela/main.py]: Reenabled option.

File: stimela/backends/kubernetes/__init__.py (hunk 3 above)
Comment (o-smirnov) [reply stimela/backends/kubernetes/__init__.py:6]: Let's not be PEP-8 nazis, nothing wrong with compounding standard library imports when there's many to be done.

File: stimela/backends/kubernetes/run_kube.py (hunk 4 above)
Comment (o-smirnov) [reply stimela/backends/kubernetes/run_kube.py:2]: See above.

**Comment by o-smirnov** [comment]
@SpheMakh recheck please. Along with https://github.com/caracal-pipeline/cult-cargo/pull/3

**Review by SpheMakh — APPROVED** [review]

**Pull request merged**